### PR TITLE
Create volume from snapshot

### PR DESCRIPTION
**Pull request by BaiHuoYu**
**What this PR does / why we need it**:
This PR adds support to create volume from snapshot in the CSI driver.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
Implements create volume from snapshot for CSI driver.
```

### CODE REVIEW

File: csi/server/plugin/opensds/controller_test.go [hunk 1]
```diff
@@ -320,3 +339,49 @@ func TestListSnapshots(t *testing.T) {
...
+			Type: &csi.VolumeContentSource_Snapshot{
+				Snapshot: &csi.VolumeContentSource_SnapshotSource{
+					Id: "3769855c-a102-11e7-b772-17b880d2f537",
+				},
+			},
+		},
+	}
+
+	rs, err := fakePlugin.CreateVolume(fakeCtx, &fakeReq)
+
```
Comment (leonwanghui) [inline csi/server/plugin/opensds/controller_test.go]: Please remove this blank line

**Comment by xing-yang** [comment]
Please update the deployment yaml to use a new external-provisioner image.  Use mine until the official one is out:

    image: xyang105/csi:csi-provisioner
    imagePullPolicy: IfNotPresent

Also add a new example yaml file to create volume from snapshot.

**Comment by xing-yang** [comment]
Add a release note.

**Comment by xing-yang** [comment]
@BaiHuoYu Were you able to test this using the external-provisioner image I gave you?  Does it work now?

**Comment by BaiHuoYu** [comment]
> @BaiHuoYu Were you able to test this using the external-provisioner image I gave you? Does it work now?
The creation of the volume from the snapshot was successful, but the nginx pod failed.

**Comment by xing-yang** [comment]
The nginx pod failure is caused by something else, not related to snapshot functionality at all.  There're some config changes you need to make to work around that.  I'll find an updated work around and let you know.

File: csi/server/deploy/kubernetes/csi-provisioner-rbac.yaml [hunk 2]
```diff
@@ -27,8 +27,21 @@ rules:
   - apiGroups: [""]
     resources: ["events"]
     verbs: ["list", "watch", "create", "update", "patch"]
-
-    
+  - apiGroups: [""]
+    resources: ["endpoints"]
+    verbs: [ "create", "delete", "get", "list","watch","update"]
```
Comment (xing-yang) [inline csi/server/deploy/kubernetes/csi-provisioner-rbac.yaml]: Actually we only need "get" and "update".

File: csi/server/deploy/kubernetes/csi-provisioner-rbac.yaml [hunk 3]
```diff
@@ -27,8 +27,21 @@ rules:
...
+  - apiGroups: [""]
+    resources: ["endpoints"]
```
Comment (xing-yang) [inline csi/server/deploy/kubernetes/csi-provisioner-rbac.yaml]: This is a duplicate of line 31.

**Comment by xing-yang** [comment]
Please address the comments.  I'll merge it after that.  Thanks.

**Comment by BaiHuoYu** [comment]
> Please address the comments. I'll merge it after that. Thanks.

Already modified

**Pull request merged**